### PR TITLE
Add constructors for typed arrays in DOMMatrix and DOMMatrixReadOnly

### LINKED ISSUE #20349: Implement DOMMatrixReadOnly and DOMMatrix constructors for typed arrays
Now that #20267 has merged we are able to implement the fromFloat32Array and fromFloat64Array static methods, along with toFloat32Array and toFloat64Array.

Spec: https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-fromfloat32array
Code: `components/script/dom/webidls/DOMMatrixReadOnly.webidl`, `components/script/dom/dommatrixreadonly.rs`, `components/script/dom/webidls/DOMMatrix.webidl`, `components/script/dom/dommatrix.rs`
Tests: `./mach test-wpt tests/wpt/web-platform-tests/css/geometry` (there should be new [passing tests](https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations))

### PR DESCRIPTION
**Pull request by pvdrz**
- DOMMatrix now have the functions: FromFloat32Array, FromFloat64Array.
- DOMMatrixReadOnly now have the functions: FromFloat32Array, FromFloat64Array, ToFloat32Array, ToFloat64Array.

r? @jdm 

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #20349 (github issue number if applicable).

- [X] There are tests for these changes OR
- [ ] These changes do not require tests because _____

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/20405)

### CODE REVIEW

File: components/script/dom/dommatrix.rs [hunk 1]
```diff
@@ -56,315 +57,331 @@ impl DOMMatrix {
     pub fn from_readonly(global: &GlobalScope, ro: &DOMMatrixReadOnly) -> DomRoot<Self> {
         Self::new(global, ro.is_2d(), ro.matrix().clone())
     }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrix-fromfloat32array
+    #[allow(unsafe_code)]
+    pub fn FromFloat32Array(global: &GlobalScope, mut array: CustomAutoRooterGuard<Float32Array>) ->
+        Fallible<DomRoot<DOMMatrix>> {
```
Comment (jdm) [inline components/script/dom/dommatrix.rs]: Let's format these new methods like:
```rust
pub fn FromFloat32Array(
    global: &GlobalScope,
    mut array: CustomAutoRooterGuard<Float32Array>)
-> Fallible<DomRoot<DOMMatrix>> {
```

File: components/script/dom/dommatrix.rs [hunk 2]
```diff
@@ -56,315 +57,331 @@ impl DOMMatrix {
...
+        let vec: Vec<f64> = unsafe { array.as_slice() }.iter().map(|&x| x as f64).collect();
+        DOMMatrix::Constructor_(global, vec)
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrix-fromfloat64array
+    #[allow(unsafe_code)]
+    pub fn FromFloat64Array(global: &GlobalScope, mut array: CustomAutoRooterGuard<Float64Array>) ->
+        Fallible<DomRoot<DOMMatrix>> {
+            let vec: Vec<f64> = unsafe { array.as_slice() }.to_vec();
```
Comment (jdm) [inline components/script/dom/dommatrix.rs]: These can use the safe `array.to_vec()` APIs that were recently added.

File: components/script/dom/dommatrixreadonly.rs [hunk 3]
```diff
@@ -334,6 +340,23 @@ impl DOMMatrixReadOnly {
...
+    #[allow(unsafe_code)]
+    pub fn FromFloat32Array(global: &GlobalScope, mut array: CustomAutoRooterGuard<Float32Array>) ->
+        Fallible<DomRoot<DOMMatrixReadOnly>> {
+        let vec: Vec<f64> = unsafe { array.as_slice() }.iter().map(|&x| x as f64).collect();
+        DOMMatrixReadOnly::Constructor_(global, vec)
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-fromfloat64array
+    #[allow(unsafe_code)]
+    pub fn FromFloat64Array(global: &GlobalScope, mut array: CustomAutoRooterGuard<Float64Array>) ->
```
Comment (jdm) [inline components/script/dom/dommatrixreadonly.rs]: Same comments as the other method.

File: components/script/dom/dommatrixreadonly.rs [hunk 4]
```diff
@@ -553,6 +576,25 @@ impl DOMMatrixReadOnlyMethods for DOMMatrixReadOnly {
...
         DOMPoint::new(&self.global(), x, y, z, w)
     }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-tofloat32array
+    #[allow(unsafe_code)]
+    unsafe fn ToFloat32Array(&self, cx: *mut JSContext) -> NonNull<JSObject> {
+        let vec: Vec<f32> = self.matrix
+            .borrow().to_row_major_array().iter().map(|&x| x as f32).collect();
+        rooted!(in (cx) let mut array = ptr::null_mut::<JSObject>());
+        let _ = Float32Array::create(cx, CreateWith::Slice(&vec), array.handle_mut());
```
Comment (jdm) [inline components/script/dom/dommatrixreadonly.rs]: We should call unwrap() here, or we could end up with a null pointer.

File: components/script/dom/dommatrixreadonly.rs [hunk 5]
```diff
@@ -553,6 +576,25 @@ impl DOMMatrixReadOnlyMethods for DOMMatrixReadOnly {
...
+        NonNull::new_unchecked(array.get())
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-fromfloat64array
+    #[allow(unsafe_code)]
+    unsafe fn ToFloat64Array(&self, cx: *mut JSContext) -> NonNull<JSObject> {
+        let arr = self.matrix.borrow().to_row_major_array();
+        rooted!(in (cx) let mut array = ptr::null_mut::<JSObject>());
+        let _ = Float64Array::create(cx, CreateWith::Slice(&arr), array.handle_mut());
```
Comment (jdm) [inline components/script/dom/dommatrixreadonly.rs]: Unwrap.

File: components/script/dom/dommatrixreadonly.rs [hunk 6]
```diff
@@ -553,6 +576,25 @@ impl DOMMatrixReadOnlyMethods for DOMMatrixReadOnly {
...
+    #[allow(unsafe_code)]
+    unsafe fn ToFloat32Array(&self, cx: *mut JSContext) -> NonNull<JSObject> {
+        let vec: Vec<f32> = self.matrix
+            .borrow().to_row_major_array().iter().map(|&x| x as f32).collect();
+        rooted!(in (cx) let mut array = ptr::null_mut::<JSObject>());
+        let _ = Float32Array::create(cx, CreateWith::Slice(&vec), array.handle_mut());
+        NonNull::new_unchecked(array.get())
+    }
+
+    // https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-fromfloat64array
```
Comment (jdm) [inline components/script/dom/dommatrixreadonly.rs]: https://drafts.fxtf.org/geometry-1/#dom-dommatrixreadonly-tofloat64array

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/dommatrix.rs (hunk 1 above)
Comment (christianpoveda) [reply components/script/dom/dommatrix.rs]: ok, will fix :)

File: components/script/dom/dommatrix.rs (hunk 2 above)
Comment (christianpoveda) [reply components/script/dom/dommatrix.rs]: ohh ok ok that's good

File: components/script/dom/dommatrixreadonly.rs (hunk 4 above)
Comment (christianpoveda) [reply components/script/dom/dommatrixreadonly.rs]: ok will fix

**Pull request merged**